### PR TITLE
Use a single transaction for fsck, it is much faster.

### PR DESCRIPTION
**Pull request by andrewchambers**
40 minute wall time down to 43 seconds. My benchmark was a filesystem which had a linux kernel file tree extracted into it, then hard aborted.

Before:
```
2019-02-20_19:20:02.28825 Using cached metadata.
2019-02-20_19:20:02.40102 WARNING: Remote metadata is outdated.
2019-02-20_19:20:02.40111 Checking DB integrity...
2019-02-20_19:20:08.22732 Creating temporary extra indices...
2019-02-20_19:20:08.36617 Checking lost+found...
2019-02-20_19:20:08.37370 Checking for dirty cache objects...
..processed 4040/4040 files (100%)..Checking names (refcounts)...
2019-02-20_19:23:48.66639 Checking contents (names)...
2019-02-20_19:23:48.68062 Checking contents (inodes)...
2019-02-20_19:23:48.69212 Checking contents (parent inodes)...
2019-02-20_19:23:48.70171 Checking objects (reference counts)...
2019-02-20_19:23:48.89381 Checking objects (backend)...
2019-02-20_19:24:18.59634 Checking objects (sizes)...
2019-02-20_19:24:18.60110 Checking blocks (referenced objects)...
2019-02-20_19:28:55.66042 Checking blocks (refcounts)...
2019-02-20_19:28:55.71518 Checking blocks (checksums)...
2019-02-20_19:28:55.71962 Checking inode-block mapping (blocks)...
2019-02-20_19:39:25.05754 Checking inode-block mapping (inodes)...
2019-02-20_19:39:25.06671 Checking inodes (refcounts)...
2019-02-20_19:39:25.12355 Checking inodes (sizes)...
2019-02-20_19:47:03.59538 Checking extended attributes (names)...
2019-02-20_19:47:03.59560 Checking extended attributes (inodes)...
2019-02-20_19:47:03.59572 Checking symlinks (inodes)...
2019-02-20_19:47:03.59586 Checking directory reachability...
2019-02-20_19:47:20.28490 Checking unix conventions...
2019-02-20_19:47:21.04894 Checking referential integrity...
2019-02-20_20:02:54.93460 Dropping temporary indices...
2019-02-20_20:02:54.94057 Dumping metadata...
```
After:
```
2019-02-20_21:44:05.66374 Using cached metadata.
2019-02-20_21:44:05.81843 WARNING: Remote metadata is outdated.
2019-02-20_21:44:05.81852 Checking DB integrity...
2019-02-20_21:44:11.52812 Checking lost+found...
2019-02-20_21:44:11.53598 Checking for dirty cache objects...
..processed 4040/4040 files (100%)..Checking names (refcounts)...
2019-02-20_21:44:20.40988 Checking contents (names)...
2019-02-20_21:44:20.42321 Checking contents (inodes)...
2019-02-20_21:44:20.43239 Checking contents (parent inodes)...
2019-02-20_21:44:20.44007 Checking objects (reference counts)...
2019-02-20_21:44:20.49055 Checking objects (backend)...
2019-02-20_21:44:48.20491 Checking objects (sizes)...
2019-02-20_21:44:48.20951 Checking blocks (referenced objects)...
2019-02-20_21:44:48.21633 Checking blocks (refcounts)...
2019-02-20_21:44:48.26585 Checking blocks (checksums)...
2019-02-20_21:44:48.26600 Checking inode-block mapping (blocks)...
2019-02-20_21:44:48.27407 Checking inode-block mapping (inodes)...
2019-02-20_21:44:48.28250 Checking inodes (refcounts)...
2019-02-20_21:44:48.33719 Checking inodes (sizes)...
2019-02-20_21:44:48.39712 Checking extended attributes (names)...
2019-02-20_21:44:48.39729 Checking extended attributes (inodes)...
2019-02-20_21:44:48.39739 Checking symlinks (inodes)...
2019-02-20_21:44:48.39749 Checking directory reachability...
2019-02-20_21:44:48.48049 Checking unix conventions...
2019-02-20_21:44:48.86817 Checking referential integrity...
2019-02-20_21:44:48.91469 Dropping temporary indices...
2019-02-20_21:44:48.92328 Dumping metadata...
```

### CODE REVIEW

**Comment by andrewchambers** [comment]
I noticed some operations in fs.py do the same thing, I wonder if doing the same for some of those would be beneficial.

File: src/s3ql/fsck.py [hunk 1]
```diff
@@ -110,10 +110,14 @@ def check(self, check_cache=True):
             self.check_loops()
             self.check_unix()
             self.check_foreign_keys()
+        except:
+            self.conn.execute('ROLLBACK')
```
Comment (andrewchambers) [inline src/s3ql/fsck.py]: Hmm, I just noticed journal is off by default, so this won't work...

**Comment by Nikratio** [comment]
This is odd. SQLite is configured with 

```
           'PRAGMA synchronous = OFF',
           'PRAGMA journal_mode = OFF',
```

which I think should turn `BEGIN` and `END` into no-ops. Which SQLite version are you using?

**Comment by Nikratio** [comment]
Also, if I read the log output correctly, there were only read operations on the database (excluding creation of the indices), which shouldn't be affected that much by transactions either. Are you sure that you tested this correctly? Can you reproduce the speed-up?

**Comment by andrewchambers** [comment]
I will run the experiment again to double check.

**Comment by andrewchambers** [comment]
Closing until I can confirm, it may have been an IO issue with my vm.

**Pull request closed without merging**